src/ripple/ledger/impl/ApplyView.cpp

```diff
@@ -0,0 +1,298 @@
+
+    if (! root)
+    {
+        // No root, make it.
+        root = std::make_shared<SLE>(directory);
+        root->setFieldH256 (sfRootIndex, directory.key);
+        describe (root);
+
+        STVector256 v;
+        v.push_back (key);
+        root->setFieldV256 (sfIndexes, v);
+
+        insert (root);
+
+        return { true, 0 };
+    }
+
+    std::uint64_t page = root->getFieldU64(sfIndexPrevious);
+
+    auto node = root;
+
+    if (page)
+    {
+        node = peek (keylet::page(directory, page));
+
+        if (!node)
+            LogicError ("Directory chain: root back-pointer broken.");
+    }
+
+    auto indexes = node->getFieldV256(sfIndexes);
+
+    // If there's space, we use it:
+    if (indexes.size () < dirNodeMaxEntries)
+    {
+        // If we're not maintaining strict order, then
+        // sort entries lexicographically.
+        if (!strictOrder)
+        {
+            auto v = indexes.value();
+
+            // We can't use std::lower_bound here because
+            // existing pages may not be sorted
+            auto pos = std::find_if(v.begin(), v.end(),
+                [&key](uint256 const& item)
+                {
+                    return key < item;
+                });
+
+            if (pos != v.end() && key == *pos)
+                LogicError ("dirInsert: double insertion");
+
+            v.insert (pos, key);
+            indexes = v;
+        }
+        else
+            indexes.push_back(key);
+
+        node->setFieldV256 (sfIndexes, indexes);
+        update(node);
+        return { true, page };
+    }
+
+    // Check whether we're out of pages.
+    if (++page == 0)
+        return { false, 0 };
+
+    // Insert the new key:
+    indexes.clear();
+    indexes.push_back (key);
+
+    // We are about to create a new node; we'll link it to
+    // the chain first:
+
+    node->setFieldU64 (sfIndexNext, page);
+    update(node);
+
+    root->setFieldU64 (sfIndexPrevious, page);
+    update(root);
+
+    node = std::make_shared<SLE>(keylet::page(directory, page));
+    node->setFieldH256 (sfRootIndex, directory.key);
+    node->setFieldV256 (sfIndexes, indexes);
+
+    // Save some space by not specifying the value 0 since
+    // it's the default.
+    if (page != 1)
+        node->setFieldU64 (sfIndexPrevious, page - 1);
+    describe (node);
+    insert (node);
+
+    return { true, page };
+}
+
+std::pair<bool, std::uint64_t>
+ApplyView::dirInsert (
+    Keylet const& directory,
+    Keylet const& key,
+    bool strictOrder,
+    std::function<void (std::shared_ptr<SLE> const&)> describe)
+{
+    return dirInsert (directory, key.key, strictOrder, describe);
+}
+
+bool
+ApplyView::dirRemove (
+    Keylet const& directory,
+    std::uint64_t currPage,
+    uint256 const& key,
+    bool keepRoot)
+{
+    SLE::pointer node = peek(keylet::page(directory, currPage));
+
+    if (!node)
+        return false;
+
+    {
+        auto entries = node->getFieldV256(sfIndexes);
```

Comment:
It does. We can get away with it with a bit of ugliness:

    auto entries = dynamic_cast<STVector256&>(node->getField(sfIndexes));
    if (!entries)
        LogicError ("Invalid sfIndexes type ");